src/DeviceInterface/InterfacePrototype.js

```diff
@@ -254,16 +321,121 @@ class InterfacePrototype {
+            topContextData.credentials = [fromPassword(password)]
+        }
+
+        this.attachTooltipInner(form, input, getPosition, click, topContextData)
     }
 
-    attachTooltipInner (form, input, inputType, getPosition, _click) {
+    /**
+     * Create a new password once for a given HTMLFormElement
+     *
+     * @param {HTMLFormElement} formElement
+     * @param {HTMLInputElement} input
+     * @returns {string}
+     */
+    createAndStorePassword (formElement, input) {
+        const prev = this.generatedPasswords.get(formElement)
+        if (prev) return prev
+        const newPassword = generate({
+            input: input.getAttribute('passwordrules'),
+            domain: window.location.hostname
+        })
+        this.generatedPasswords.set(formElement, newPassword)
+        return newPassword
+    }
+
+    /**
+     * If the device was capable of generating password, and it
+     * previously did so for the form in question, then offer to
+     * save the credentials
+     *
+     * @param {{ formElement?: HTMLFormElement; }} options
+     */
+    shouldPromptToStoreCredentials (options) {
+        if (!options.formElement) return false
+        if (!this.supportsFeature('password.generation')) return false
+
+        // if we previously generated a password, allow it to be saved
+        if (this.generatedPasswords.has(options.formElement)) {
+            return true
+        }
+
+        return false
+    }
+
+    /**
+     * When an item was selected, we then call back to the device
+     * to fetch the full suite of data needed to complete the autofill
+     *
+     * @param {InputTypeConfigs} config
+     * @param {(CreditCardObject|IdentityObject|CredentialsObject)[]} items
+     * @param {string|number} id
+     */
+    onSelect (config, items, id) {
+        id = String(id)
+        const matchingData = items.find(item => String(item.id) === id)
+        if (!matchingData) throw new Error('unreachable (fatal)')
+
+        const dataPromise = (() => {
+            switch (config.type) {
+            case 'creditCard': return this.getAutofillCreditCard(id)
+            case 'identities': return this.getAutofillIdentity(id)
+            case 'credentials': {
+                if (id === GENERATED_ID) {
+                    return Promise.resolve({ success: matchingData })
+                }
+                return this.getAutofillCredentials(id)
+            }
+            default: throw new Error('unreachable!')
+            }
+        })()
+
+        // wait for the data back from the device
+        dataPromise.then(response => {
+            if (response.success) {
+                return this.selectedDetail(response.success, config.type)
+            } else {
+                return Promise.reject(new Error('none-success response'))
+            }
+        }).catch(e => {
+            console.error(e)
+            return this.removeTooltip()
```

Comment:
Does `selectedDetail` also call `removeTooltip`? Or how is the tooltip closed in that case? Is it worth placing `removeTooltip` in a `finally`?

Reply:
removal of the tooltip is already handled elsewhere, yes. So the `this.removeTooltip()` in this `catch` is only there to ensure we can recover.